sympy/codegen/ast.py

```diff
@@ -399,3 +448,515 @@ def body(self):
         Must be an iterable object or CodeBlock.
         """
         return self._args[2]
+
+
+class Token(Basic):
+    """ Similar to Symbol, but takes no assumptions.
+
+    Defining fields are set in __slots__.
+    """
+
+    __slots__ = []
+
+    def __new__(cls, *args, **kwargs):
+        if len(args) == 1 and not kwargs and isinstance(args[0], cls):
+            return args[0]
+        args = args + tuple([kwargs[k] for k in cls.__slots__[len(args):]])
+        return Type.__xnew_cached_(cls, *args)
+
+    def __new_stage2__(cls, *args):
+        obj = Basic.__new__(cls)
+        for attr, arg in zip(cls.__slots__, args):
+            setattr(obj, attr, arg)
+        return obj
+
+    __xnew_cached_ = staticmethod(cacheit(__new_stage2__))
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        for attr in self.__slots__:
+            if getattr(self, attr) != getattr(other, attr):
+                return False
+        return True
+
+    def __hash__(self):
+        return hash(tuple([getattr(self, attr) for attr in self.__slots__]))
```

Comment:
Oh right. I think you have to define `__hash__` and have it call `super(Type, self).__hash__`. There are other examples in the codebase.